### ISSUE TITLE
Please consider adding type annotations to networkx

### ISSUE THREAD

**Issue by NeilGirdhar** [networkx/networkx#3988]
Type annotations helps with linting and discovering bugs early.  They also act as documentation.

**Comment by ericmjl** [comment]
@NeilGirdhar I think that would be a great suite of PRs! As long as @dschult agrees, I wouldn't mind helping you review them if you're open to kickstarting the type annotations. (Btw, doesn't have to be the entire library in one PR, one or a handful of files at a time would probably be good.)

**Comment by dschult** [comment]
Thanks for the suggestion here.  This seems more complex than it first appears. For example, there is no Type Node.  Indeed, nodes can be any hashable object. Similarly, there is no type for an edge. We currently document types in docstrings for parameters and for return values. But those types are often not standard. For example:  string, callable, bool or None    
Do we lose this kind of flexibility with the annotations?

I can imagine this ballooning into a huge maintenance headache. Can you help me to understand the costs of such a change? Are there automated ways (like Travis-CI) to check what is annotated and are there automated ways to manage the collection of types we use?

**Comment by NeilGirdhar** [comment]
> This seems more complex than it first appears. For example, there is no Type Node. 

Yes, that's why I used a `TypeVar`, and made the graph a generic class.  That means that users can use it with any type, but only if they decide to restrict the type (by requesting `DiGraph[str]` for example), are the node types restricted.  `DiGraph[Hashable]` or just `DiGraph` would maintain the current type freedom.

Your point about nodes being hashable is well-taken, so I should have written:
```python
Node = TypeVar('Node', bound=Hashable)
```

> We currently document types in docstrings for parameters and for return values. But those types are often not standard. For example: string, callable, bool or None
> Do we lose this kind of flexibility with the annotations?

That's great!  That will make the work of type-annotations much easier.  The benefit of annotations is that linters like MyPy can check user code.  FYI, those types you mentioned are specified as `str`, `Callable[[ParameterTypes...], ReturnType]`, `bool`, and `None`.  It's very simple.

> I can imagine this ballooning into a huge maintenance headache. Can you help me to understand the costs of such a change? Are there automated ways (like Travis-CI) to check what is annotated and are there automated ways to manage the collection of types we use?

`mypy --disallow-untyped-calls --disallow-untyped-defs --disallow-incomplete-defs` reports all the missing type annotations if you choose to see them.

The rest of the Python world is moving towards type annotations as it's a real luxury to have it library code.  There's something satisfying about type-checking your code and seeing it pass, a feeling of security.  I've found many bugs in my code using `mypy`.

That said, I understand your reluctance to avoid maintenance.  I remember an old comment of yours "Can we get back to writing algorithms?"  😄   I understand.  You don't have to go crazy annotating everything if you don't want to.  I wasn't planning on helping since I have way too much work, but I guess I could do `Graph` and `DiGraph` to get you started.

Also, you're already maintaining the docstrings, and this would replace that work.  So, in a sense it doesn't seem like more work than you're already doing.  And the type annotations are more useful.

**Comment by NeilGirdhar** [comment]
This is what type annotations might look like: https://github.com/networkx/networkx/pull/4014

**Comment by fmagin** [comment]
There was also a previous discussion about this in https://github.com/networkx/networkx/issues/2183 which I haven't seen mentioned yet.

I am also quite interested in type annotations for networkx and would be willing to work on this too. I started a `networkx-stub` package at https://github.com/fmagin/networkx-stubs/ which at least one person has already started extending with the intent to cover MultiGraphs. I can also vouch for the usefulness of type annotations in complex code that uses `networkx`, mostly in the context of program analysis. My main question is where the effort should be focused, i.e. is the sentiment still that the annotations belong in a separate `networkx-stubs` package like discussed in the previous issue, or is supporting Python 2 no longer a concern anymore, i.e. inline annotations are welcome?
If the latter is the case there is still a decision to be made what Python version should at least be supported. The newer versions bring quite some useful features e.g. literal types [PEP 586](https://www.python.org/dev/peps/pep-0586/) requiring 3.8, class and instance variable annotations [PEP 526](https://www.python.org/dev/peps/pep-0526/#class-and-instance-variable-annotations) requiring 3.6.
My experience has been that 3.6 is a reasonable tradeoff between supporting older version and leveraging useful newer features, but networkx is a far larger project than others I have worked on so the requirements here might very well be different.

Concerning the questions above:
> I can imagine this ballooning into a huge maintenance headache. Can you help me to understand the costs of such a change? Are there automated ways (like Travis-CI) to check what is annotated and are there automated ways to manage the collection of types we use?

With the right mypy parameters you could also only check for inconsistent annotations. Code that is added after the initial annotations and typing support could still be untyped as long as it doesn't violate any existing annotations. This would either imply an issue with the new code or the existing annotations, both something that should be caught by the CI.

**Comment by dschult** [comment]
Sorry for poking my head in here again when the discussion is really over in #4014 but I want to ask something related to a comment above.

Is there a way to specify that a Node should be "hashable and not None"?

**Comment by fmagin** [comment]
According to https://stackoverflow.com/a/38078003/13220684 there is [typing.Hashable](https://docs.python.org/3/library/typing.html#typing.Hashable). `None` is only ever allowed if using `Optional`. So just `typing.Hashable` should be enough as far as I understand.

**Comment by NeilGirdhar** [comment]
@fmagin You got it, `typing.Hashable` is the right way.

In my pull request, I did want to make sure that Node was hashable, which is why I did:
```
Node = TypeVar('Node', bound=Hashable)
```
That means that `Node` must be a subtype of `Hashable`.

Technically `None` should be hashable, but [according to mypy, it's not](https://github.com/python/mypy/issues/8768).  So, I guess you have the behavior you want.

**Comment by danieleades** [comment]
i've just started using this library, and i strongly agree with this. Usage would be far more discoverable with type annotations.

One slight speedbump with adding stricter mypy config to this library is that the `src` and `tests` are not separated. Often you'll want a stricter mypy config for source than for tests (at least, while incrementally adding annotations), which is painful to do when there's no separation. Is there a reason for not having separate directories?

**Comment by a-r-j** [comment]
Hi folks, wanted to share a small library I made for this purpose: https://github.com/a-r-j/graphtype

If there’s interest I’d be happy to help build this in to NetworkX

**Comment by songololo** [comment]
@a-r-j worth notifying people on this discussion as well:
https://github.com/networkx/networkx/pull/4014

**Comment by danieleades** [comment]
> Hi folks, wanted to share a small library I made for typehinting (and checking!) attributed graphs: https://github.com/a-r-j/graphtype
> 
> If there’s interest I’d be happy to help build this in to NetworkX

as you can probably see from the discussion over at #4014 the real challenge here is in convincing maintainers that type annotations are a welcome improvement to the library.

**Comment by a-r-j** [comment]
@danieleades 

I see, it's certainly a lively discussion 😁 

I think this suggestion flies slightly under that radar - it seems that the concerns are primarily about readability of nx source code. I think that the feature I suggest is primarily for the benefit of users building on top of nx. If they don't like annotations they simply don't have to use them. Whilst I am in favour of annotations more broadly as they improve readability imo, this feature doesn't have to add any internal annotations if the maintainers are not convinced.

**Comment by zwelz3** [comment]
Looks like someone is working on stubs independent of official sources
https://github.com/eggplants/networkx-stubs/tree/master/networkx-stubs

Maybe consider a PR to bring these in if they work with 3.0? (Haven't had a chance to test)

**Comment by danieleades** [comment]
they look possibly auto-generated? lots of `${arg}: Any` in there.
certainly a lot less accurate and less useful (at least, at this stage) than what was proposed in #4014 

maybe the author can comment? (@eggplants)

**Comment by eggplants** [comment]
@danieleades networkx-stubs is a previously generated stubgen file. It is not maintained now, but I would like to update it if necessary.

**Comment by NeilGirdhar** [comment]
@eggplants You're very welcome to lift everything from my pull request into the stubs if you want !

**Comment by danieleades** [comment]
any movement on this?

The [adoption rate of type annotations in python libraries is extremely high](https://py-code.org/stats) and i feel the maintainers of this library are *very* much on the wrong side of history here...

**Comment by fmagin** [comment]
Thanks to @rhelmot my stubs have now been polished and [merged into typeshed](https://github.com/python/typeshed/pull/10544) and I will direct my future efforts there, until I get a very clear signal from the networkx team that they are willing to integrate in-line annotations.

One problem is that 3.2 now introduced a new code pattern that makes type annotating much more verbose and complicated https://github.com/python/typeshed/pull/10544#issuecomment-1776036652
I haven't looked into this, but this seems like a general pattern with networkx code: It is often written in a way that makes it fairly complicated to annotate with correct types. This basically always means that the exact behavior of the code is hard to reason about too, which is IMO a bad sign if it is widespread in a codebase that doesn't have very good reasons for this. But I haven't looked into this new change enough to form an opinion on the benefits of the change though.

I think the overall direction is starting to become clear: There is currently enough demand for and value in at least basic type annotations such that the wider community will create and maintain them, even without the cooperation of the networkx team itself. This demand is only growing, and I expect that more and more people will start to see the benefits.

I think this topic and discussion hinges on what networkx is trying to be: Should networkx be a library for mathematicians and researchers for graphs as the mathematical objects and their study? Or a library for computer scientists, software engineers and programmers that concerns itself with graphs as the widely useful data-structure as part of larger projects? Ideally it could be both, but currently it seems that it is focusing more on the former use-case (or a different one I am missing?), but due to ecosystem inertia, it is still widely used for the software use-case. But I'm personally starting to become motivated to switch to a library that has better support for type integrity checks and better runtime and memory performance.

**Comment by NeilGirdhar** [comment]
@rhelmot @fmagin This is awesome!!  I was wondering where the type annotations were coming from in the edition of PyRIght.  Thank you very much for taking initiative and getting this in!

I rebased https://github.com/networkx/networkx/pull/4014, and I'm going to look into merging some of the changes I have in it that are not yet in the typeshed, and also I'm getting an errors too that I'll submit a PR for.

Regarding adding annotations to NetworkX, with time, I've come to understand that there are a variety of people who simply fell in love with Python before type annotations and just don't see them the way us "typed Python people" do.  Even if we disagree, I think it's important to respect that NetworkX is "someone else's baby".  We should be careful about pushing too hard.  In a way, we are after all guests here.

That said, for anyone who missed it, I believe this was @dschult's [most recent comment](https://github.com/networkx/networkx/pull/4014#issuecomment-1190746022) on the topic.  In summary, the @dschult feels that annotated code is less readable, but he likes the benefit of automatic IDE checking that annotations give, and he supports adding stubs to NetworkX itself.

He also expressed some ideas about the future of coding: "I expect that it will become much easier to see the code in the form that you want to see it".  This is something I totally agree with in general.  I would love to see code in my style, but for it to be automatically committed in the project style (black, Google-python-style, etc.).  Dan wants this for automatic merging splitting of stubs from code so that anti-type people could see code without types, and pro-type people could see it with types).  A great idea, but I don't think anything like that is on the horizon.  So I don't think we can realistically wait for that.

He also responded to my PR, which did more than just annotate stubs.  The PR tried to make NetworkX itself pass type checking, and so it involved some changes to the code (LBYL over EAFP).  In retrospect, it was a mistake to do this in the first annotation PR.  I had thought that being able to run the type checker would be motivating for the NetworkX team.  I think these changes only alienated them.

He also talked about docstring annotations and tools to keep them consistent with ordinary annotations. So he's also resistant to removing the docstring annotations.

All that said, I'm interested to know if there's been any change on the NetworkX's team feelings about adding inline annotations since their last comment a year and half ago.  Do they still feel that they make code less readable?  If so, we can continue to update the typeshed or maybe add stubs to NetworkX.

**Comment by danieleades** [comment]
> I'm personally starting to become motivated to switch to a library that has better support for type integrity checks and better runtime and memory performance.

I feel like the best outcome would be a small, highly-performant, and fully-annotated graph structure library, and then a larger, separate library implementing a rich collection of algorithms on top of that. Basically I would propose that `networkx` split out the core graph data structures into a separate library for general use, and then continue tinkering away in a separate library that imports it as a dependency.

In the meantime, i'm going to take a look at [rustworkx](https://github.com/Qiskit/rustworkx) (formally retworkx). it still uses out of source annotations, but at least they're in the same repository. Looks like they're possibly autogenerated from Rust bindings.

**Comment by rossbar** [comment]
Thank you @fmagin for your nice summary, I think your take on the situation largely aligns with my own.

I'll go ahead and restate my stance here: I'm personally +1 for adding stub files (`.pyi`) directly to this repository. This enables static type checking, which is currently not available in NetworkX.

> There is a genuine trade-off here, because if we put the stub files into the networkx repository as separate files, this still puts the burden of reviewing PRs on the networkx team itself.

This is the main concern from my perspective. Like all OSS libraries, NetworkX is review-bandwidth-limited. There isn't nearly enough review for the PRs that actually deal with the functionality of the library itself. The review consideration is also a big part of the motivation (for me personally at least) to use stub files instead of inline annotations. That way, all of the typing information is self-contained, and the review pattern can essentially be: "this PR only touches `.pyi` files, so we can be more aggressive about merging". This is the approach that was taken by NumPy (another community I participate in) and IMO it worked very well.

> I'm doing this because I have the gut feeling that this is the better distribution of effort across the ecosystem, as long as the networkx project would only accept stubs.

I'm not fully aware of how typeshed works, but this sounds like a good approach - anything that helps provide review specifically for annotations sounds good to me. Again, I'm +1 on adding stubs directly to networkx itself as well.

> I think some of us were pushing for types in networkx in a bit too strong way, and I think it comes down to this conceptual mixup between a graph library and a software project built for research on graphs. 

For my part, I wouldn't way that anything is necessarily "too strong", but these conversations in the past have always seemed to stall on the specifics. Type annotations have been discussed many times in community meetings and it's generally something networkx would like to support. However, as has been stated many times before, the developers *do* have an opinion on an adoption plan that the community is most comfortable with - i.e. starting with stub files instead of inline annotations. In the past, this has been met with a lot of push back from those who seem most interested in type annotations (i.e. the participants of these conversations) so we've been at an impasse and generally haven't been able to push forward.

**Comment by NeilGirdhar** [comment]
I think the stubs being in the typeshed is probably an excellent outcome for now.

There is a big difference in development time.  In the typeshed,
* The review process is really fast (<2 days for my last changes), and
* The changes are available in the next type checker build (<7 days for pyright typically).

Whereas, we would otherwise be looking at NetworkX's release cycle for new annotations, which could be months.

So, for now while the stubs are worked on, the typeshed seems ideal.  Once they're mature, we can talk about moving the stubs into NetworkX, which would mainly be useful if the NetworkX maintainers:
* want to oversee what goes into the stubs, or
* want to add types to a new feature in the same pull request as that feature (rather than having to file a parallel pull request for the stubs).

**Comment by eriknw** [comment]
> One problem is that 3.2 now introduced a new code pattern that makes type annotating much more verbose and complicated [python/typeshed#10544 (comment)](https://github.com/python/typeshed/pull/10544#issuecomment-1776036652)

I created #7185 as a possible way to address this to allow typeshed to continue using standard Python function definitions. Is there an easy way (such as with a decorator) in typeshed to add types for function attributes?

Does anybody here want to learn what `_dispatch` is all about?

**Comment by rhelmot** [comment]
That's a very interesting pull request. I'll attempt to redo types on top of HEAD when I have time. I am also interested in learning about _dispatch.

**Comment by NeilGirdhar** [comment]
> To support that syntax, the `Graph` class only needs to define `__class_getitem__` method. Inline annotations is irrelevant.

Right, exactly.  That's why I was suggesting inheriting from `Generic[Node]` to provide that.  Would such a PR be likely to be accepted?

**Comment by eriknw** [comment]
> I am also interested in learning about _dispatch.

Great! `_dispatch` is related to a new, experimental feature in NetworkX that allows for additional backend implementations of graph data structures and algorithms. It is strictly opt-in for now. This is meant to address one of the biggest pain points of NetworkX users: performance! The goal is to make it easy for users to accelerate and scale their code if NetworkX is too slow.

I know of three NetworkX backends that are being worked on:
- `nx-cugraph`: run super-fast on GPU
  - https://github.com/rapidsai/cugraph/tree/branch-24.02/python/nx-cugraph
- `graphblas-algorithms`: run in parallel on the CPU with OpenMP
  - https://github.com/python-graphblas/graphblas-algorithms
- `nx-parallel`: https://github.com/networkx/nx-parallel (in progress)

The cuGraph and GraphBLAS backends in particular often have state-of-the-art performance (or close to it) for algorithms they implement, and efficient binary data structures as long as the graphs aren't frequently mutating.

We want to make life easier for backend implementers:
- backends can run NetworkX tests
- backends can show up in NetworkX online documentation
  - for example, scroll to bottom of:
    - https://networkx.org/documentation/latest/reference/algorithms/generated/networkx.algorithms.link_analysis.pagerank_alg.pagerank.html
- backends can be explicitly used from NetworkX:
  - when calling an algorithm, explicitly choose a backend; graph `G` will be converted
    - `nx.pagerank(G, backend="graphblas")`
  - when _creating_ a graph, choose a backend:
    - `G = nx.from_pandas_edgelist(df, backend="cugraph")`
  - automatically (with configuration--and we plan to make this better)
- plus some other TLC to make the experience more pleasant

There's still a lot we plan to do (especially better documentation!), hence why dispatching is still experimental. For example, I want "data structure"-only backends to be easy to create and generally useful. For example, it may be nice to use dict-of-arrays using COO or CSR or DCSR to store edge indices (similar to what `nx-cugraph` does), or maybe use scipy sparse arrays, or maybe pandas DataFrames, or maybe igraph graphs, or rustworkx graphs, etc., and then have backend-to-backend conversions.

Adding dispatching to networkx algorithms has the added benefit of indicating what edge, node, or graph data is used by the function, and what the default values are For example:
- `pagerank`, uses "weight" edge data (with default value of 1, which is the default): https://github.com/networkx/networkx/blob/b7d0b0cc8f7395dc9668d25379f3615752112c20/networkx/algorithms/link_analysis/pagerank_alg.py#L9-L10
- `connected_components` doesn't use any edge or node data--only graph structure: https://github.com/networkx/networkx/blob/b7d0b0cc8f7395dc9668d25379f3615752112c20/networkx/algorithms/components/connected.py#L16-L17
- `min_cost_flow_cost` uses "demand" node data and two edge attributes with different default values: https://github.com/networkx/networkx/blob/b7d0b0cc8f7395dc9668d25379f3615752112c20/networkx/algorithms/flow/mincost.py#L10-L11
- `k_truss` uses _all_ attributes (node, edge, and graph): https://github.com/networkx/networkx/blob/b7d0b0cc8f7395dc9668d25379f3615752112c20/networkx/algorithms/core.py#L471-L472

Hence, I think this is in the same vein as typing in that it's declaring how a graph will be used and what attributes it needs. This effort has uncovered a number of (minor) bugs in NetworkX. As we implement more algorithms in backends, I'm curious whether we'll find some awkward functions that are also awkward for typing.

That's the gist of it. NetworkX is evolving to better support binary data structures and alternative implementations, and will keep its current data structures. I'm happy to answer any questions.

**Comment by eriknw** [comment]
> > To support that syntax, the `Graph` class only needs to define `__class_getitem__` method. Inline annotations is irrelevant.
> 
> Right, exactly. That's why I was suggesting inheriting from `Generic[Node]` to provide that. (I'm not suggesting annotating any methods or functions.) Would such a PR be likely to be accepted?

I'm not a core dev, so I'm not in a position to say or guess. I'm not an expert on typing, but there don't appear to be any usability or performance penalties that I can find by adding this. Are there any caveats that we should be aware of?

**Comment by NeilGirdhar** [comment]
> Are there any caveats that we should be aware of?

Just that we would be fixing the number of generic parameters to one: the `Node` type.  It seems very unlikely to change that though.

**Comment by eriknw** [comment]
> Just that we would be fixing the number of generic parameters to one: the `Node` type. It seems very unlikely to change that though.

This does seem to be the most natural way to begin. One could imagine using [`pandera`-like](https://pandera.readthedocs.io/en/stable/index.html) typing with a schema for edge, node, and graph attributes and their types, which may become more appealing once backends with binary data structures become more commonplace.

**Comment by rossbar** [comment]
Official types for NetworkX are available on [`typeshed`](https://github.com/python/typeshed/tree/main/stubs/networkx) and can be installed (explicitly) with `pip install types-networkx` and are automatically grabbed by type-checking tooling. The NetworkX developers are incredibly grateful for `typeshed` and all the folks who contribute to maintain the networkx type annotations!
